lib/AST/DiagnosticConsumer.cpp

```diff
@@ -165,29 +165,57 @@ FileSpecificDiagnosticConsumer::subconsumerForLocation(SourceManager &SM,
 void FileSpecificDiagnosticConsumer::handleDiagnostic(
     SourceManager &SM, SourceLoc Loc, DiagnosticKind Kind,
     StringRef FormatString, ArrayRef<DiagnosticArgument> FormatArgs,
-    const DiagnosticInfo &Info) {
+    const DiagnosticInfo &Info, const SourceLoc defaultDiagnosticLoc) {
 
   HasAnErrorBeenConsumed |= Kind == DiagnosticKind::Error;
 
-  Optional<FileSpecificDiagnosticConsumer::Subconsumer *> subconsumer;
+  auto subconsumer =
+      findSubconsumerForAnyKind(SM, Loc, Kind, defaultDiagnosticLoc);
+  if (subconsumer) {
+    subconsumer.getValue()->handleDiagnostic(
+        SM, Loc, Kind, FormatString, FormatArgs, Info, defaultDiagnosticLoc);
+    return;
+  }
+  // Last resort: spray it everywhere
+  for (auto &subconsumer : Subconsumers)
+    subconsumer.handleDiagnostic(SM, Loc, Kind, FormatString, FormatArgs, Info,
+                                 defaultDiagnosticLoc);
+}
+
+Optional<FileSpecificDiagnosticConsumer::Subconsumer *>
+FileSpecificDiagnosticConsumer::findSubconsumerForAnyKind(
```

Comment:
I'd just drop the "ForAnyKind" here. This is the general entry point for finding a consumer; the other one specifically calls out the "non-note"-ness. I guess it is edging a little close to the lower-level `subconsumerForLocation`, but the longer name isn't addressing that part anyway.

Reply:
Sure! I tried thinking of a name including "AndMakingSureThatNotesWouldBeInTheSamePlaceAsDiagnostics" but I couldn't find anything I liked.